### PR TITLE
Add include of generated files

### PR DESCRIPTION
**Pull request by bobot**
This branch tries to add features needed for #3901 . Here my first difficulties:

  - The source tree is context independent, e.g what is a vendored directory doesn't depend on the `Context.t`.
     -  I think it is a good thing, but it means that generated included files should not include `(vendored ...)` stanza since the file is generated in a context
  - The parsing of all the dune files is done in one call `Dune_load.load` for all the directories. So the generation of the included files creates a cycles with all the dune files, so with the inclusion of their content. Two possibilities:
     -  Separate the loading of the dune files by directory.
     -  Include the files later, for example when the rules are evaluated, it could restrict even more the possible stanza but it would change less things.

Currently instead of extending the `(include file)` stanza I added `(include_generated file)` in order to simplify the developement of the feature. And for the user it helps to separate the two cases.

### CODE REVIEW

**Comment by rgrinberg** [comment]
I'm going to dissent against having a separate stanza. There's a couple of reasons why I don't like it:

* We aren't offering anything new conceptually, only removing an existing limitation. Why should we have 2 separate stanzas for the same construct in the language?
* It is a bad UI for end users. Now they'll need to churn their dune files from include generated sources.

> Perhaps even guard the feature under a (using include_generated 0.1) in the dune-project file? So that we don't have to commit to stability and versioning too early.

This can be done without a separate stanza. It seems simple to guard the feature behind a `Path.Source.is_in_source_tree` check when it's disabled.

**Comment by jeremiedimino** [comment]
Eventually, we should break down `Super_context` BTW. This thing is ugly and pre-dates `Memo`.

File: CHANGES.md [hunk 1]
```diff
@@ -1,6 +1,9 @@
 Unreleased
 ----------
 
+- Add experimental include of generated file in dune files (#5009, fixes
```
Comment (christinerose) [inline CHANGES.md:4]:
```suggestion
- Add experimental include of generated file in Dune files (#5009, fixes
```

**Review by christinerose — CHANGES_REQUESTED** [review]

**Comment by jeremiedimino** [comment]
> @rgrinberg @jeremiedimino Source_tree doesn't depend on the context (or build_context), right?

That's right.

> So it seems that adding the support of scoped_dirs which depends on the build_context since the generation of the file depends on it requires more changes if done in Source_tree. Do you have some advice?

The new feature I'm working on should help (see my comment on slack). In the meantime, I suggest to focus on scoped dirs that don't depend on the build context.

**Comment by jeremiedimino** [comment]
@christinerose we decided to abandon the feature in its current form, so it's probably not worth reviewing the current PR.

**Comment by jeremiedimino** [comment]
@bobot why do you want the `scoped_dir` status to depend on the build context BTW? Whether a directory is a scoped directory or not seems more like a universal property to me.

> EDIT: Why does Build_system depends on Source_tree?

For various reasons. For instance, some internal behaviours of Dune are tied to the `(lang dune x.y)` version, and this version is obtained via the source tree. Though it doesn't have to: it could be returned by `gen_rules` or another callback.

**Comment by jeremiedimino** [comment]
That makes sense. To be fair, it does feel like we are stuffing `Source_tree` too much. Things like the vendored status or cram tests don't have much to do with `Source_tree`. Maybe even the `dune-project`. All this could be part of `src/dune_rules`.

**Comment by christinerose** [comment]
> @christinerose we decided to abandon the feature in its current form, so it's probably not worth reviewing the current PR.

Should I dismiss the pending reviews then, or is this for new changes?

**Comment by christinerose** [comment]
Great, @bobot. Thanks!

**Comment by rgrinberg** [comment]
@bobot can you rebase? I'd like to take another look at this PR. I think it should be somewhat cleaner now that we've removed unmemoized file access

**Comment by Alizter** [comment]
@bobot ping, are you planning to make progress with this?

**Pull request closed without merging**